tools/transform.cpp

```diff
@@ -301,8 +301,39 @@ TransformVerify::TransformVerify(Transform &t, bool check_each_var) :
       tgt_instrs.emplace(i.getName(), &i);
     }
   }
+}
+
+Errors TransformVerify::sync() {
+  try {
+    t.tgt.syncDataWithSrc(t.src);
+
+    // Check sizes of global variables
+    auto globals_tgt = t.tgt.getGlobalVars();
+    auto globals_src = t.src.getGlobalVars();
+    for (auto GVS: globals_src) {
+      auto I = find_if(globals_tgt.begin(), globals_tgt.end(),
+        [GVS](auto *GV) -> bool { return GVS->getName() == GV->getName(); });
+      if (I == globals_tgt.end())
+        continue;
 
-  t.tgt.syncDataWithSrc(t.src);
+      auto GVT = *I;
+      if (GVS->size() != GVT->size()) {
+        stringstream ss;
+        ss << "Global variable " << GVS->getName() << " in source and "
+          << "target has different sizes (" << GVS->size() << ", "
+          << GVT->size() << ")";
+        throw AliveException(ss.str(), false);
+      } else if (GVS->isConst() != GVT->isConst()) {
+        stringstream ss;
+        ss << "Global variable " << GVS->getName() << " in source and "
+          << "target has different constness";
```

Comment:
Maybe give a more precise message by case splitting, e.g.:
Unsupported transformation: Global variable foo is constant in target but not in source.

The reverse transformation (const->non const) should yield an unsound error (it's a bug!).

Reply:
This involves changing `blks_readonly` because the address is currently shared between source and target - having different constness of same global variable in src and tgt will make precondition always false.
Do you prefer having a separate PR that relaxes the condition to nonconst -> const, or address it in this PR?

Reply:
The bug https://bugs.llvm.org/show_bug.cgi?id=43616 was successfully found because previously `blocks_readonly` was an array.
I think I can make `blks_readonly` be defined in source and target separately, so there are two `blks_readonly` UFs.

Reply:
Non-const -> Const: we can't verify this
Const -> Non-const: the transformation  is wrong.

We should distinguish these cases.